corehq/ex-submodules/couchexport/export.py

```diff
@@ -313,19 +322,24 @@ def get_data(self):
     @classmethod
     def wrap_all_rows(cls, tables):
         """
-        Take a list of tuples (name, SINGLE_ROW) or (name, (ROW, ROW, ...))
+        Take a list of tuples (name, SINGLE_ROW) or (name, (ROW, ROW, ...)). The rows can be generators.
         """
-        ret = []
-        for name, rows in tables:
-            rows = list(rows)
-            if rows and (not hasattr(rows[0], '__iter__') or isinstance(rows[0], six.string_types)):
-                soft_assert_type_text(rows[0])
+
+        def _wrap_row(first_entry, rows):
+            if first_entry and (not hasattr(first_entry, '__iter__') or isinstance(first_entry, six.string_types)):
+                soft_assert_type_text(first_entry)
                 # `rows` is actually just a single row, so wrap it
                 rows = [rows]
-            ret.append(
-                (name, [cls(row) for row in rows])
-            )
-        return ret
+            return rows
+
+        wrapped_rows = []
+        for name, rows in tables:
+            rows = itertools.chain([], rows)
+            first_entry_rows, rows_to_wrap = itertools.tee(rows)
+            first_entry = next(first_entry_rows)
+            wrapped_rows.append(
+                (name, [cls(row) for row in _wrap_row(first_entry, rows_to_wrap)]))
+        return wrapped_rows
```

Comment:
I'm still not crazy about this method.  I find it a little hard to follow.  I like your idea of extracting `_wrap_row`, but I'd pull the `first_entry` business into it too, since that's the only place it's needed.  I'd also avoid the usage of `tee`, since it [stores the list in memory when needed](https://docs.python.org/2/library/itertools.html#itertools.tee), (although the compiler may be smart enough to see that `first_entry_rows` isn't accessed again, not sure).  Lastly, you can also use a generator expression rather than a list expression and avoid evaluating the rows here.

I realize some of those suggestions are kind of vague, so here's my attempt:

```python
def coerce_to_list_of_rows(rows):
    rows = iter(rows)
    # peek at the first element, then add it back
    first_entry = next(rows)
    rows = itertools.chain([first_entry], rows)
    if first_entry and (not hasattr(first_entry, '__iter__') or isinstance(first_entry, six.string_types)):
        soft_assert_type_text(first_entry)
        # `rows` is actually just a single row, so wrap it
        return [rows]
    return rows

return [
    (name, (cls(row) for row in coerce_to_list_of_rows(rows)))
    for name, rows in tables
]
```

Reply:
https://github.com/dimagi/commcare-hq/pull/23782/commits/a2ff0cb8e39658671d727969d4ba6302c76ba343